scipy/stats/stats.py

```diff
@@ -4153,6 +4154,170 @@ def binary_search(n, n1, n2, side):
+    the 15 subjects (47%) became infected. The data are tabulated as a 2 × 2
+    table::
+
+                    Vaccine  Placebo
+        Yes             8        2
+        No              1        5
+
+    We use this table to find the p-value:
+    >>> import scipy.stats as stats
+    >>> TX, pvalue = stats.barnard_exact([[7, 12], [8, 3]])
+    >>> pvalue
+    0.0681...
+
+    """
+    if nuisance_num < 0:
+        raise ValueError(
+            "Number of nuisance's samples must be non-negative, "
+            f"found {nuisance_num!r}"
+        )
+
+    c = np.asarray(
+        table,
+        dtype=np.int64)
+
+    if not c.shape == (2, 2):
+        raise ValueError("The input `table` must be of shape (2, 2).")
+
+    if np.any(c < 0):
+        raise ValueError("All values in `table` must be nonnegative.")
+
+    if 0 in c.sum(axis=0) or 0 in c.sum(axis=1):
+        # If both values in a row or column are zero, the p-value is 1 and
+        # the score's statistic is NaN.
+        return np.nan, 1.0
+
+    total_c1, total_c2 = c.sum(axis=0)
+    n = total_c1 + total_c2
+
+    x1 = np.arange(total_c1 + 1, dtype=np.int64).reshape(-1, 1)
+    x2 = np.arange(total_c2 + 1, dtype=np.int64).reshape(1, -1)
+
+    p1, p2 = x1 / total_c1, x2 / total_c2
+
+    if variance == "pooled":
+        p = (x1 + x2) / (total_c1 + total_c2)
+        var_p1_p2 = p * (1 - p) * (1 / total_c1 + 1 / total_c2)
+    elif variance == "unpooled":
+        var_p1_p2 = p1 * (1 - p1) / total_c1 + p2 * (1 - p2) / total_c2
+    else:
+        msg = (
+            "`variance` should be one of {{'pooled', 'unpooled'}}, "
+            f"found {variance!r}"
+        )
+        raise ValueError(msg)
+
+    # To avoid warning when dividing by 0
+    with np.errstate(divide='ignore', invalid='ignore'):
+        TX = (p2 - p1) / np.sqrt(var_p1_p2)
+
+    TX_obs = TX[
+        c[0, 0], c[0, 1]
+    ]
+
+    if alternative == "two-sided":
+        TX = np.abs(TX)
+        idx = TX >= abs(TX_obs)
+    elif alternative == "less":
+        idx = TX <= TX_obs
+    elif alternative == "greater":
+        idx = TX >= TX_obs
+    else:
+        msg = "`alternative` should be one of {'two-sided', 'less', 'greater'},"
+        f" found {alternative!r}"
+        raise ValueError(msg)
+
+    # Pass x1 and x2 in dimension 3
+    x1 = x1.reshape(-1, 1, 1)
+    x2 = x2.reshape(1, -1, 1)
+
+    nuisance_arr = np.linspace(
+        start=0, stop=1, num=nuisance_num + 2)
+    nuisance_arr = nuisance_arr[1:-1]  # Remove 0 and 1
+    # Reshape in dimension 3 array
+    nuisance_arr = nuisance_arr.reshape(1, 1, -1)
+
+    x1_comb = _barnard_exact_compute_gammaln_combination(total_c1)
+    x2_comb = _barnard_exact_compute_gammaln_combination(total_c2)
+
+    PX = np.exp(x1_comb[x1] + x2_comb[x2] + np.log(nuisance_arr) * (
+        x1 + x2) + np.log(1 - nuisance_arr) * (n - x1 - x2))
+    p_value = PX[idx].sum(axis=0).max()  # Just sum where TX >= TX0
+    # and take the max value
```

Comment:
The optimization strategy (now a grid search) should be improved. E.g. a iterative grid search.

Reply:
BEFORE OPTIMISATION : 
%timeit tx_obs, res = stats.barnard_exact([[100, 2], [1000, 5]])
102 ms ± 857 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)

AFTER : 
%timeit tx_obs, res = barnard_exact([[100, 2], [1000, 5]])
83 ms ± 1.36 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

Better optimization for better result 👍 🥇

Reply:
But still not able to compute large table : 
```
tx_obs, res = stats.barnard_exact([[14500, 20000], [30000, 40000]], nuisance_num=30)
Killed
```

Reply:
Large table means exploration of large space of combination. Here the number of combination to explore is (14500+30000)×(20000+40000) approximatively 2.7×10⁹. With 30 values, and 8 octets (a float) by values, this means around 600 GiB in memory.

In fact, the Barnard test is not useful in this case.